### ISSUE TITLE
Stateful Layer: New state gets new dimension attached?

### ISSUE THREAD

**Issue by Artur-Galstyan** [patrick-kidger/equinox#580]
Hey Patrick,

this one is a bit tricky to explain, but here's the MVC:

```python
import functools as ft

import equinox as eqx
import jax.numpy as jnp
from icecream import ic
from jaxtyping import Array


class Counter(eqx.nn.StatefulLayer):
    index: eqx.nn.StateIndex

    def __init__(self):
        init_state = 0
        self.index = eqx.nn.StateIndex(init_state)

    def __call__(
        self, x: Array, state: eqx.nn.State, *, key
    ) -> tuple[Array, eqx.nn.State]:
        value = state.get(self.index)
        ic(x.shape, value.shape)
        new_x = x + value
        new_state = state.set(self.index, value + 1)
        ic(new_state)
        return new_x, new_state


class OuterModel(eqx.Module):
    counters: eqx.nn.Sequential

    def __init__(self):
        # My use case is a transformer and the "Counter" emulates a single "Block" (which contains 
        # the MHA implementation with KV cache, which is the state)
        self.counters = eqx.nn.Sequential([Counter() for _ in range(3)])

    def __call__(
        self, x: Array, state: eqx.nn.State, *, key, **kwargs
    ) -> tuple[Array, eqx.nn.State]:
        x, state = self.counters(x, state)

        return x, state


def main():
    model, state = eqx.nn.make_with_state(OuterModel)()
    ic(state)
    x = jnp.ones(shape=(2, 3))

    partial_model = ft.partial(model, state=state, key=None)
    output, state = eqx.filter_vmap(partial_model)(x)
    ic(output, state)

    partial_model = ft.partial(model, state=state, key=None)
    output, state = eqx.filter_vmap(partial_model)(x)
    ic(state)


if __name__ == "__main__":
    main()
```

Here's the output:

```
ic| state: State(0x101c86e98=i32[], 0x101c86eb8=i32[], 0x101c86ed8=i32[])
ic| x.shape: (3,), value.shape: ()
ic| new_state: State(0x101c86e98=i32[], 0x101c86eb8=i32[], 0x101c86ed8=i32[])
ic| x.shape: (3,), value.shape: ()
ic| new_state: State(0x101c86e98=i32[], 0x101c86eb8=i32[], 0x101c86ed8=i32[])
ic| x.shape: (3,), value.shape: ()
ic| new_state: State(0x101c86e98=i32[], 0x101c86eb8=i32[], 0x101c86ed8=i32[])
ic| output: Array([[1., 1., 1.],
                   [1., 1., 1.]], dtype=float32)
    state: State(0x101c86e98=i32[2], 0x101c86eb8=i32[2], 0x101c86ed8=i32[2])
ic| x.shape: (3,), value.shape: (2,)
```

I was dabbling in KV caches and noticed that my KV matrix shapes changed from

`(max_seq_len, num_heads, n_embd)` to `(1, max_seq_len, num_heads, n_embd)`

For some reason, an additional dimension gets added to the matrices. I managed to create an MVC for this. 

I'm not sure if this is a bug or intended. I saw this [other example in the docs](https://docs.kidger.site/equinox/api/nn/stateful/#vmapd-stateful-layers) but couldn't quite make it work:

```python

...
# this does not work!
def main():
    model, state = eqx.nn.make_with_state(OuterModel)()
    ic(state)
    x = jnp.ones(shape=(2, 3))

    model = functools.partial(model, key=None)
    v_model = eqx.filter_vmap(model, axis_size=2)
    substate = state.substate(v_model)

    output, substate = v_model(x, substate)
    ic(output, substate)
    state = state.update(substate)

    substate = state.substate(v_model)
    output, substate = v_model(x, substate)
    ic(output, state)
    state = state.update(substate)
```

In the `vmapd-stateful-layers` example, the `vmap` happens "inside" the layer, but in my case, it's "outside". Long story short: how can I prevent the creation of a new dimension?

(If I just squeeze `axis=0`, it complains to me that the state's structure has changed :( )

**Comment by patrick-kidger** [comment]
If I understand your issue correctly, then the extra dimension appears on this line:
```python
output, state = eqx.filter_vmap(partial_model)(x)
```
for which the output of `filter_vmap` will have a batch dimension. You probably want `filter_vmap(..., out_axes=(0, None))` to control this.

**Comment by Artur-Galstyan** [comment]
That did indeed solve it :D Thanks!
